gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoMessageRequestFormatter.java

```diff
@@ -0,0 +1,120 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.core.BetaApi;
+import com.google.api.pathtemplate.PathTemplate;
+import com.google.protobuf.Message;
+import java.util.List;
+import java.util.Map;
+
+/** Creates parts of a HTTP request from a protobuf message. */
+@BetaApi
+public class ProtoMessageRequestFormatter<RequestT extends Message>
+    implements HttpRequestFormatter<RequestT> {
+
+  private final FieldsExtractor<RequestT, String> requestBodyExtractor;
+  private final FieldsExtractor<RequestT, Map<String, List<String>>> queryParamsExtractor;
+  private final PathTemplate pathTemplate;
+  private final FieldsExtractor<RequestT, Map<String, String>> pathVarsExtractor;
+
+  private ProtoMessageRequestFormatter(
+      FieldsExtractor<RequestT, String> requestBodyExtractor,
+      FieldsExtractor<RequestT, Map<String, List<String>>> queryParamsExtractor,
+      PathTemplate pathTemplate,
+      FieldsExtractor<RequestT, Map<String, String>> pathVarsExtractor) {
+    this.requestBodyExtractor = requestBodyExtractor;
+    this.queryParamsExtractor = queryParamsExtractor;
+    this.pathTemplate = pathTemplate;
+    this.pathVarsExtractor = pathVarsExtractor;
+  }
+
+  public static <RequestT extends Message>
+      ProtoMessageRequestFormatter.Builder<RequestT> newBuilder() {
+    return new Builder<>();
+  }
+
+  /* {@inheritDoc} */
+  @Override
+  public Map<String, List<String>> getQueryParamNames(RequestT apiMessage) {
```

Comment:
I don't understand why a method called  getQueryParamNames is returning a map? Names would seem to be a list.

Reply:
It is implementing an existing interface. As I  understand it, key is a name of a parameter, List is its values (to support array query parameters):
```
?param1=val2,val3
    ^         ^ 
String key   List<String> map value
```